File: mobile/lib/features/marketplace/models/marketplace_models.dart
import 'package:flutter/foundation.dart';

enum SellerStatus { pendingApproval, approved, suspended, rejected }
enum OfferStatus { active, paused, outOfStock, underReview }
enum FulfillmentType { fulfilledByMilterra, sellerDirect }
enum DealType { dealOfTheDay, lightningDeal, festivalSpecial }
enum CouponType { percentage, flat }
enum AuditAction { priceChange, imageUpdate, stockAdjust, dealCreate, sellerApproval, sellerSuspension, catalogCreate, statusChange }

@immutable
class SellerAccount {
  const SellerAccount({
    required this.id,
    required this.businessName,
    this.tradeName,
    this.gstin,
    this.fssaiLicense,
    this.contactEmail,
    this.contactPhone,
    this.warehouseCity,
    this.warehouseState,
    this.bankAccountNumber,
    this.ifscCode,
    this.upiId,
    this.status = SellerStatus.pendingApproval,
    this.commissionRatePercent = 8.0,
    this.ratingScore = 4.8,
    this.totalRatingsCount = 0,
    required this.createdAt,
  });

  final String id;
  final String businessName;
  final String? tradeName;
  final String? gstin;
  final String? fssaiLicense;
  final String? contactEmail;
  final String? contactPhone;
  final String? warehouseCity;
  final String? warehouseState;
  final String? bankAccountNumber;
  final String? ifscCode;
  final String? upiId;
  final SellerStatus status;
  final double commissionRatePercent;
  final double ratingScore;
  final int totalRatingsCount;
  final DateTime createdAt;

  SellerAccount copyWith({
    SellerStatus? status,
    double? commissionRatePercent,
    double? ratingScore,
    int? totalRatingsCount,
  }) =>
      SellerAccount(
        id: id,
        businessName: businessName,
        tradeName: tradeName,
        gstin: gstin,
        fssaiLicense: fssaiLicense,
        contactEmail: contactEmail,
        contactPhone: contactPhone,
        warehouseCity: warehouseCity,
        warehouseState: warehouseState,
        bankAccountNumber: bankAccountNumber,
        ifscCode: ifscCode,
        upiId: upiId,
        status: status ?? this.status,
        commissionRatePercent: commissionRatePercent ?? this.commissionRatePercent,
        ratingScore: ratingScore ?? this.ratingScore,
        totalRatingsCount: totalRatingsCount ?? this.totalRatingsCount,
        createdAt: createdAt,
      );
}

@immutable
class SellerOffer {
  const SellerOffer({
    required this.id,
    required this.productId,
    required this.sellerId,
    required this.sellerName,
    required this.sellerSku,
    required this.mrp,
    required this.sellingPrice,
    this.discountPercent = 0.0,
    required this.availableStock,
    this.lowStockThreshold = 5,
    this.deliveryPromise = 'FREE delivery by Tomorrow',
    this.fulfillmentType = FulfillmentType.fulfilledByMilterra,
    this.offerStatus = OfferStatus.active,
    this.sellerRating = 4.8,
    this.isBuyBoxWinner = false,
  });

  final String id;
  final String productId;
  final String sellerId;
  final String sellerName;
  final String sellerSku;
  final double mrp;
  final double sellingPrice;
  final double discountPercent;
  final int availableStock;
  final int lowStockThreshold;
  final String deliveryPromise;
  final FulfillmentType fulfillmentType;
  final OfferStatus offerStatus;
  final double sellerRating;
  final bool isBuyBoxWinner;

  SellerOffer copyWith({
    double? sellingPrice,
    double? mrp,
    double? discountPercent,
    int? availableStock,
    int? lowStockThreshold,
    OfferStatus? offerStatus,
    bool? isBuyBoxWinner,
  }) =>
      SellerOffer(
        id: id,
        productId: productId,
        sellerId: sellerId,
        sellerName: sellerName,
        sellerSku: sellerSku,
        mrp: mrp ?? this.mrp,
        sellingPrice: sellingPrice ?? this.sellingPrice,
        discountPercent: discountPercent ?? this.discountPercent,
        availableStock: availableStock ?? this.availableStock,
        lowStockThreshold: lowStockThreshold ?? this.lowStockThreshold,
        deliveryPromise: deliveryPromise,
        fulfillmentType: fulfillmentType,
        offerStatus: offerStatus ?? this.offerStatus,
        sellerRating: sellerRating,
        isBuyBoxWinner: isBuyBoxWinner ?? this.isBuyBoxWinner,
      );
}

@immutable
class DealPromotion {
  const DealPromotion({
    required this.id,
    required this.title,
    required this.dealType,
    required this.productId,
    required this.productTitle,
    required this.productImage,
    required this.mrp,
    required this.dealPrice,
    required this.discountPercent,
    required this.startTime,
    required this.endTime,
    this.isActive = true,
  });

  final String id;
  final String title;
  final DealType dealType;
  final String productId;
  final String productTitle;
  final String productImage;
  final double mrp;
  final double dealPrice;
  final double discountPercent;
  final DateTime startTime;
  final DateTime endTime;
  final bool isActive;
}

@immutable
class PlatformCoupon {
  const PlatformCoupon({
    required this.id,
    required this.code,
    required this.description,
    this.discountType = CouponType.percentage,
    required this.discountValue,
    this.minOrderValue = 0.0,
    this.maxDiscountCap,
    this.validUntil,
    this.usageCount = 0,
    this.isActive = true,
  });

  final String id;
  final String code;
  final String description;
  final CouponType discountType;
  final double discountValue;
  final double minOrderValue;
  final double? maxDiscountCap;
  final DateTime? validUntil;
  final int usageCount;
  final bool isActive;
}

@immutable
class MarketplaceAuditLog {
  const MarketplaceAuditLog({
    required this.id,
    required this.userRole,
    required this.userIdentifier,
    required this.action,
    required this.entityType,
    required this.entityId,
    required this.details,
    required this.timestamp,
  });

  final String id;
  final String userRole;
  final String userIdentifier;
  final AuditAction action;
  final String entityType;
  final String entityId;
  final String details;
  final DateTime timestamp;
}

enum ListingCategory { cow, buffalo, calf, heifer, bull, goat, sheep, equipment, feed }
enum ListingStatus { active, sold, paused, expired }

@immutable
class MarketplaceListing {
  const MarketplaceListing({
    required this.id,
    required this.sellerId,
    this.cattleId,
    required this.category,
    required this.title,
    required this.price,
    required this.status,
    required this.createdAt,
    this.breed,
    this.ageMonths,
    this.weightKg,
    this.milkYieldLitres,
    this.fatPct,
    this.lactationNumber,
    this.isPregnant = false,
    this.monthsPregnant,
    this.healthVerified = false,
    this.vaccinationVerified = false,
    this.description,
    this.isNegotiable = true,
    this.photos = const [],
    this.locationVillage,
    this.locationDistrict,
    this.locationState,
    this.distanceKm,
    this.viewsCount = 0,
    this.inquiriesCount = 0,
    this.sellerName,
    this.sellerPhone,
    this.isFavorited = false,
  });

  final String id, sellerId, title;
  final String? cattleId,
      breed,
      description,
      locationVillage,
      locationDistrict,
      locationState,
      sellerName,
      sellerPhone;
  final ListingCategory category;
  final ListingStatus status;
  final double price;
  final int? ageMonths, lactationNumber, monthsPregnant;
  final double? weightKg, milkYieldLitres, fatPct, distanceKm;
  final bool isPregnant,
      healthVerified,
      vaccinationVerified,
      isNegotiable,
      isFavorited;
  final List<String> photos;
  final int viewsCount, inquiriesCount;
  final DateTime createdAt;

  factory MarketplaceListing.fromJson(Map<String, dynamic> j) =>
      MarketplaceListing(
        id: j['id'].toString(),
        sellerId: j['seller_id'].toString(),
        cattleId: j['cattle_id']?.toString(),
        category: ListingCategory.values.byName(j['category'] as String? ?? 'cow'),
        title: j['title'] as String? ?? '',
        price: (j['price'] as num?)?.toDouble() ?? 0,
        status:
            ListingStatus.values.byName(j['status'] as String? ?? 'active'),
        createdAt: DateTime.tryParse(j['created_at'] as String? ?? '') ?? DateTime.now(),
        breed: j['breed'] as String?,
        ageMonths: j['age_months'] as int?,
        weightKg: (j['weight_kg'] as num?)?.toDouble(),
        milkYieldLitres: (j['milk_yield_litres'] as num?)?.toDouble(),
        fatPct: (j['fat_pct'] as num?)?.toDouble(),
        lactationNumber: j['lactation_number'] as int?,
        isPregnant: j['is_pregnant'] as bool? ?? false,
        monthsPregnant: j['months_pregnant'] as int?,
        healthVerified: j['health_verified'] as bool? ?? false,
        vaccinationVerified: j['vaccination_verified'] as bool? ?? false,
        description: j['description'] as String?,
        isNegotiable: j['is_negotiable'] as bool? ?? true,
        photos:
            (j['photos'] as List? ?? []).map((x) => x.toString()).toList(),
        locationVillage: j['location_village'] as String?,
        locationDistrict: j['location_district'] as String?,
        locationState: j['location_state'] as String?,
        distanceKm: (j['distance_km'] as num?)?.toDouble(),
        viewsCount: j['views_count'] as int? ?? 0,
        inquiriesCount: j['inquiries_count'] as int? ?? 0,
        sellerName: j['seller_name'] as String?,
        sellerPhone: j['seller_phone'] as String?,
        isFavorited: j['is_favorited'] as bool? ?? false,
      );

  MarketplaceListing copyWith({bool? isFavorited}) => MarketplaceListing(
        id: id,
        sellerId: sellerId,
        cattleId: cattleId,
        category: category,
        title: title,
        price: price,
        status: status,
        createdAt: createdAt,
        breed: breed,
        ageMonths: ageMonths,
        weightKg: weightKg,
        milkYieldLitres: milkYieldLitres,
        fatPct: fatPct,
        lactationNumber: lactationNumber,
        isPregnant: isPregnant,
        monthsPregnant: monthsPregnant,
        healthVerified: healthVerified,
        vaccinationVerified: vaccinationVerified,
        description: description,
        isNegotiable: isNegotiable,
        photos: photos,
        locationVillage: locationVillage,
        locationDistrict: locationDistrict,
        locationState: locationState,
        distanceKm: distanceKm,
        viewsCount: viewsCount,
        inquiriesCount: inquiriesCount,
        sellerName: sellerName,
        sellerPhone: sellerPhone,
        isFavorited: isFavorited ?? this.isFavorited,
      );
}

@immutable
class MarketplaceFilter {
  const MarketplaceFilter({
    this.category,
    this.breed,
    this.minPrice,
    this.maxPrice,
    this.state,
    this.district,
    this.isPregnant,
    this.healthVerified,
    this.sort = 'newest',
  });

  final ListingCategory? category;
  final String? breed, state, district, sort;
  final double? minPrice, maxPrice;
  final bool? isPregnant, healthVerified;

  Map<String, dynamic> get query => {
        'category': category?.name,
        'breed': breed,
        'min_price': minPrice,
        'max_price': maxPrice,
        'state': state,
        'district': district,
        'is_pregnant': isPregnant,
        'health_verified': healthVerified,
      }..removeWhere((_, v) => v == null || v == '');
}

File: backend/app/models/marketplace_models.py
import enum
import uuid
from datetime import datetime
from decimal import Decimal
from sqlalchemy import Boolean, DateTime, ForeignKey, Integer, JSON, Numeric, String, Text, Enum as SAEnum
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column
from app.database import Base

class SellerStatus(str, enum.Enum):
    PENDING_APPROVAL = " PENDING_APPROVAL\
 APPROVED = \APPROVED\
 SUSPENDED = \SUSPENDED\
 REJECTED = \REJECTED\

class DocumentType(str, enum.Enum):
 GST_CERTIFICATE = \GST_CERTIFICATE\
 FSSAI_LICENSE = \FSSAI_LICENSE\
 PAN_CARD = \PAN_CARD\
 CANCELLED_CHEQUE = \CANCELLED_CHEQUE\

class FulfillmentType(str, enum.Enum):
 FULFILLED_BY_MILTERRA = \FULFILLED_BY_MILTERRA\
 SELLER_DIRECT = \SELLER_DIRECT\

class OfferStatus(str, enum.Enum):
 ACTIVE = \ACTIVE\
 PAUSED = \PAUSED\
 OUT_OF_STOCK = \OUT_OF_STOCK\
 MODERATION_REQUIRED = \MODERATION_REQUIRED\

class DealType(str, enum.Enum):
 DEAL_OF_THE_DAY = \DEAL_OF_THE_DAY\
 LIGHTNING_DEAL = \LIGHTNING_DEAL\
 FESTIVAL_SPECIAL = \FESTIVAL_SPECIAL\

class CouponDiscountType(str, enum.Enum):
 PERCENTAGE = \PERCENTAGE\
 FLAT = \FLAT\

class AuditActionType(str, enum.Enum):
 PRICE_CHANGE = \PRICE_CHANGE\
 IMAGE_UPDATE = \IMAGE_UPDATE\
 STOCK_ADJUST = \STOCK_ADJUST\
 DEAL_CREATE = \DEAL_CREATE\
 SELLER_APPROVAL = \SELLER_APPROVAL\
 SELLER_SUSPENSION = \SELLER_SUSPENSION\
 CATALOG_CREATE = \CATALOG_CREATE\
 STATUS_CHANGE = \STATUS_CHANGE\

class SellerAccount(Base):
 __tablename__ = \seller_accounts\

 id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
 user_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey(\users.id\), index=True, nullable=False)
 business_name: Mapped[str] = mapped_column(String(200), nullable=False)
 trade_name: Mapped[str | None] = mapped_column(String(200))
 gstin: Mapped[str | None] = mapped_column(String(20), index=True)
 fssai_license: Mapped[str | None] = mapped_column(String(30))
 contact_email: Mapped[str | None] = mapped_column(String(100))
 contact_phone: Mapped[str | None] = mapped_column(String(20))
 warehouse_city: Mapped[str | None] = mapped_column(String(100))
 warehouse_state: Mapped[str | None] = mapped_column(String(100))
 bank_account_number: Mapped[str | None] = mapped_column(String(50))
 ifsc_code: Mapped[str | None] = mapped_column(String(20))
 upi_id: Mapped[str | None] = mapped_column(String(100))
 status: Mapped[SellerStatus] = mapped_column(SAEnum(SellerStatus), default=SellerStatus.PENDING_APPROVAL, index=True)
 commission_rate_percent: Mapped[Decimal] = mapped_column(Numeric(5, 2), default=Decimal(\8.00\))
 rating_score: Mapped[Decimal] = mapped_column(Numeric(3, 2), default=Decimal(\4.80\))
 total_ratings_count: Mapped[int] = mapped_column(Integer, default=0)
 created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow, index=True)
 updated_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

class SellerDocument(Base):
 __tablename__ = \seller_documents\

 id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
 seller_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey(\seller_accounts.id\), index=True, nullable=False)
 doc_type: Mapped[DocumentType] = mapped_column(SAEnum(DocumentType), nullable=False)
 file_url: Mapped[str] = mapped_column(String(500), nullable=False)
 document_number: Mapped[str | None] = mapped_column(String(100))
 is_verified: Mapped[bool] = mapped_column(Boolean, default=False)
 created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)

class SellerOffer(Base):
 __tablename__ = \seller_offers\

 id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
 product_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey(\products.id\), index=True, nullable=False)
 seller_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey(\seller_accounts.id\), index=True, nullable=False)
 seller_sku: Mapped[str] = mapped_column(String(100), index=True, nullable=False)
 mrp: Mapped[Decimal] = mapped_column(Numeric(12, 2), nullable=False)
 selling_price: Mapped[Decimal] = mapped_column(Numeric(12, 2), nullable=False)
 discount_percent: Mapped[Decimal] = mapped_column(Numeric(5, 2), default=Decimal(\0.00\))
 available_stock: Mapped[int] = mapped_column(Integer, default=0)
 low_stock_threshold: Mapped[int] = mapped_column(Integer, default=5)
 delivery_promise_days: Mapped[int] = mapped_column(Integer, default=2)
 fulfillment_type: Mapped[FulfillmentType] = mapped_column(SAEnum(FulfillmentType), default=FulfillmentType.FULFILLED_BY_MILTERRA)
 offer_status: Mapped[OfferStatus] = mapped_column(SAEnum(OfferStatus), default=OfferStatus.ACTIVE, index=True)
 is_buy_box_winner: Mapped[bool] = mapped_column(Boolean, default=False)
 created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)
 updated_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

class DealPromotion(Base):
 __tablename__ = \deal_promotions\

 id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
 title: Mapped[str] = mapped_column(String(200), nullable=False)
 deal_type: Mapped[DealType] = mapped_column(SAEnum(DealType), default=DealType.DEAL_OF_THE_DAY)
 product_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey(\products.id\), index=True, nullable=False)
 offer_id: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), ForeignKey(\seller_offers.id\))
 deal_price: Mapped[Decimal] = mapped_column(Numeric(12, 2), nullable=False)
 discount_percent: Mapped[Decimal] = mapped_column(Numeric(5, 2), nullable=False)
 start_time: Mapped[datetime] = mapped_column(DateTime, nullable=False)
 end_time: Mapped[datetime] = mapped_column(DateTime, nullable=False)
 is_active: Mapped[bool] = mapped_column(Boolean, default=True, index=True)
 created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)

class PlatformCoupon(Base):
 __tablename__ = \platform_coupons\

 id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
 code: Mapped[str] = mapped_column(String(50), unique=True, index=True, nullable=False)
 description: Mapped[str | None] = mapped_column(String(255))
 discount_type: Mapped[CouponDiscountType] = mapped_column(SAEnum(CouponDiscountType), default=CouponDiscountType.PERCENTAGE)
 discount_value: Mapped[Decimal] = mapped_column(Numeric(10, 2), nullable=False)
 min_order_value: Mapped[Decimal] = mapped_column(Numeric(12, 2), default=Decimal(\0.00\))
 max_discount_cap: Mapped[Decimal | None] = mapped_column(Numeric(12, 2))
 valid_until: Mapped[datetime | None] = mapped_column(DateTime)
 usage_count: Mapped[int] = mapped_column(Integer, default=0)
 is_active: Mapped[bool] = mapped_column(Boolean, default=True)
 created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)

class MarketplaceAuditLog(Base):
 __tablename__ = \marketplace_audit_logs\

 id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
 user_id: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), index=True)
 user_email_or_phone: Mapped[str] = mapped_column(String(100), default=\system_admin\)
 user_role: Mapped[str] = mapped_column(String(50), default=\SUPER_ADMIN\)
 action_type: Mapped[AuditActionType] = mapped_column(SAEnum(AuditActionType), nullable=False)
 entity_type: Mapped[str] = mapped_column(String(50), nullable=False)
 entity_id: Mapped[str] = mapped_column(String(100), nullable=False)
 details: Mapped[str] = mapped_column(Text, nullable=False)
 diff_payload: Mapped[dict] = mapped_column(JSON, default=dict)
 created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow, index=True)
